### PR TITLE
Created a basic library tutorial for everyone to follow and use

### PR DESCRIPTION
**Pull request by ALsJourney**
### **PR Type**
Documentation

___

### **Description**
- Added a comprehensive tutorial on creating a Joomla library, including folder structure, key sections, and example code.
- Provided detailed instructions on packaging, installing, and using the library within Joomla.
- Introduced a new section in the documentation explaining the purpose and utility of libraries in Joomla.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>basic-library.md</strong><dd><code>Add tutorial for creating a Joomla library</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/building-extensions/libraries/basic-library.md

<li>Added a tutorial for creating a Joomla library.<br> <li> Described folder structure and key sections of the library.<br> <li> Provided example code for service class and packaging instructions.<br> <li> Included installation and usage instructions for the library.<br>

</details>

  </td>
  <td><a href="https://github.com/joomla/Manual/pull/307/files#diff-1a88f730c2084fe45b87b27d87c55e0114d4be95ad5d2c8ad2f00fbb9193dbe2">+95/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.md</strong><dd><code>Introduce libraries section in documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/building-extensions/libraries/index.md

<li>Introduced a new section on libraries.<br> <li> Explained the purpose and utility of libraries in Joomla.<br>

</details>

  </td>
  <td><a href="https://github.com/joomla/Manual/pull/307/files#diff-2e16083920e5223850344571a1d4e83aa4c1327b18525fae405f971bd722028c">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by HLeithner** [comment]
As said on Mattermost a library has no service, you don't boot a library, what are you trying to achieve? or is the word "service" meant in a different why

**Comment by ALsJourney** [comment]
> As said on Mattermost a library has no service, you don't boot a library, what are you trying to achieve? or is the word "service" meant in a different why

changed it, thanks!

File: docs/building-extensions/libraries/basic-library.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+## Usage
+
+To use the library in your Joomla extension, you need to include the library in your code. Here is an example of how to use the library:
+
+```php
+<?php
+use YourLibraryNamespace\YourLibrary;
+
+$library = new YourLibrary();
```
Comment (Fedik) [inline docs/building-extensions/libraries/basic-library.md:99]: This will throw an error `Class "YourLibraryNamespace\YourLibrary" not found`

Comment (Fedik) [reply docs/building-extensions/libraries/basic-library.md:99]: Your XML should have `<namespace path="src">VendorName\LibraryName<namespace>` element, or you have to register namespace on your own, example with plugin.

Also `YourLibrary` class need to rename to something else. E.g. `HelloWord`. Then whole thing will be:
```
use VendorName\LibraryName\HeloWord;

$hello = new HelloWord();
echo $hello->sayHello();
```

**Comment by robbiejackson** [comment]
I saved those sections of code into your suggested filenames, but the extension didn't install. 

I suggest you use "lib_example" etc instead of "yourlibraryname" as I think that would be easier to follow.

Could you also add the filenames as titles to your code sections? See https://docusaurus.io/docs/next/markdown-features/code-blocks#code-title

If you need an example, the DPMedia Free download at https://joomla.digital-peak.com/download/dpmedia/dpmedia-1.15.0 is a package which has a library within it.

Also I found out recently that the "version=" attribute in the manifest file `<extension>` tag isn't used any more, so would be best omitted. See https://github.com/joomla/joomla-cms/issues/25814 

Overall this will make a good addition to the manual - just need to get it all working correctly thanks.

File: docs/building-extensions/libraries/basic-library.md (hunk 1 above)
Comment (ALsJourney) [reply docs/building-extensions/libraries/basic-library.md]: Thank you! Added it!

**Comment by ALsJourney** [comment]
> I saved those sections of code into your suggested filenames, but the extension didn't install.
> 
> I suggest you use "lib_example" etc instead of "yourlibraryname" as I think that would be easier to follow.
> 
> Could you also add the filenames as titles to your code sections? See https://docusaurus.io/docs/next/markdown-features/code-blocks#code-title
> 
> If you need an example, the DPMedia Free download at https://joomla.digital-peak.com/download/dpmedia/dpmedia-1.15.0 is a package which has a library within it.
> 
> Also I found out recently that the "version=" attribute in the manifest file `<extension>` tag isn't used any more, so would be best omitted. See [joomla/joomla-cms#25814](https://github.com/joomla/joomla-cms/issues/25814)
> 
> Overall this will make a good addition to the manual - just need to get it all working correctly thanks.

I added a few changes. Could you try again?
If it still doesn't work, then I'm going to take a look at the links you provided.

File: docs/building-extensions/libraries/basic-library.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    <libraryname>JoomlaExampleLibrary</libraryname>
+    <author>Author Name</author>
+    <creationDate>2024-09-08</creationDate>
+    <license>GNU General Public License version 2 or later</license>
+    <version>1.0.0</version>
+    <description>Showcase on how to create a Joomla 5 library to reuse code.</description>
+    <namespace path="src">VendorName\LibraryName</namespace>
+    <files>
+        <folder>src</folder>
+        <filename>joomlaexamplelibrary.xml</filename>
```
Comment (robbiejackson) [inline docs/building-extensions/libraries/basic-library.md:45]: You need to remove this line

Comment (ALsJourney) [reply docs/building-extensions/libraries/basic-library.md:45]: which one, the namespace?
@Fedik suggested i should add it :D I'm confused now

File: docs/building-extensions/libraries/basic-library.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+2. Go to Extensions > Manage > Install.
+3. Upload the zip file containing the library.
+
+## Usage
+
+To use the library in your Joomla extension, you need to include the library in your code. Here is an example of how to use the library:
+
+```php
+<?php
+use VendorName\LibraryName\HelloWord;
```
Comment (robbiejackson) [inline docs/building-extensions/libraries/basic-library.md:98]: This should be:

use VendorName\LibraryName\YourLibrary;

File: docs/building-extensions/libraries/basic-library.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+$library = new HelloWorld();
```
Comment (robbiejackson) [inline docs/building-extensions/libraries/basic-library.md:100]: This should be 

    $library = new YourLibrary();

File: docs/building-extensions/libraries/basic-library.md (hunk 2 above)
Comment (ALsJourney) [reply docs/building-extensions/libraries/basic-library.md:45]: ok, will remove it

Comment (robbiejackson) [reply docs/building-extensions/libraries/basic-library.md:45]: No, the `<filename>` line, as that file doesn't exist.

**Comment by robbiejackson** [comment]
Yes that works fine for me now.

I suggest you leave it another day just in case there are any other comments, and then duplicate it to the other versions ready for merging.

There are a couple of references to Joomla 5 and we do copy documentation to Joomla 4.4, so that might look a bit odd there, but that sort of thing I wouldn't be overly concerned about. I didn't test the code on 4.4 but I reckon it should be ok.

**Comment by ALsJourney** [comment]
thanks for the help

**Pull request merged**